libyul/SideEffects.h

```diff
@@ -37,33 +37,87 @@ struct SideEffects
 	bool movable = true;
+	/// If true, the expressions in this code can be freely moved over other statements, expressions
+	/// or function calls, as long as these do not invalidate the blockchain state.
+	bool movableIfStateInvariant = true;
+	/// If true, the expression in this code can be freely moved over other statements, expressions
+	/// or function calls, as long as these do not invalidate the storage or the blockchain state.
+	bool movableIfStorageInvariant = true;
+	/// If true, the expressions in this code can be freely moved over other statements, expressions
+	/// or function calls, as long as these do not invalidate memory or contain a MSIZE.
+	bool movableIfMemoryInvariant = true;
 	/// If true, the code can be removed without changing the semantics.
 	bool sideEffectFree = true;
 	/// If true, the code can be removed without changing the semantics as long as
 	/// the whole program does not contain the msize instruction.
 	bool sideEffectFreeIfNoMSize = true;
-	/// If false, storage is guaranteed to be unchanged by the code under all
+	/// If false, the blockchain state is guaranteed to be unchanged by the code under all
 	/// circumstances.
+	bool invalidatesState = false;
+	/// If false, storage and the blockchain state is guaranteed to be unchanged by the code under
+	/// all circumstances.
 	bool invalidatesStorage = false;
 	/// If false, memory is guaranteed to be unchanged by the code under all
 	/// circumstances.
 	bool invalidatesMemory = false;
+	/// If true, the block will contain a MSIZE
+	bool containsMSize = false;
 
+	SideEffects() {}
+	SideEffects(
+		bool _movable,
+		bool _movableIfStateInvariant,
+		bool _movableIfStorageInvariant,
+		bool _movableIfMemoryInvariant,
+		bool _sideEffectFree,
+		bool _sideEffectFreeIfNoMSize,
+		bool _invalidatesState,
+		bool _invalidatesStorage,
+		bool _invalidatesMemory,
+		bool _containsMSize
+	):
+		movable(_movable),
+		movableIfStateInvariant(_movableIfStateInvariant),
+		movableIfStorageInvariant(_movableIfStorageInvariant),
+		movableIfMemoryInvariant(_movableIfMemoryInvariant),
+		sideEffectFree(_sideEffectFree),
+		sideEffectFreeIfNoMSize(_sideEffectFreeIfNoMSize),
+		invalidatesState(_invalidatesState),
+		invalidatesStorage(_invalidatesStorage),
+		invalidatesMemory(_invalidatesMemory),
+		containsMSize(_containsMSize)
+	{}
 	/// @returns the worst-case side effects.
 	static SideEffects worst()
 	{
-		return SideEffects{false, false, false, true, true};
+		return SideEffects{
+			false, // movable
+			false, // movableIfStateInvariant
+			false, // movableIfStorageInvariant
+			false, // movableIfMemoryInvariant
+			false, // sideEffectFree
+			false, // sideEffectFreeIfNoMSize
+			true,  // invalidatesState
+			true,  // invalidatesStorage
+			true,  // invalidatesMemory
+			true   // containsMSize
+		};
 	}
 
 	/// @returns the combined side effects of two pieces of code.
 	SideEffects operator+(SideEffects const& _other)
 	{
 		return SideEffects{
```

Comment:
I have the suspicion that we can find a more generic solution here. The fact that some "semantic information" (called `SideEffects` here) is stored negatively (combined using `&&`) and some is stored positively (combined using `||`) hints at the fact that these might be split into two structures.

Reply:
Or maybe we can split it my "affected area": `(state | storage | memory): (reads | modifies)`

Reply:
I remember having made a full list together with @ekpyron at some point, but I think we did not document it.